### PR TITLE
NVTX domains, categories, colors

### PR DESCRIPTION
**Pull request by milesial**
Spice up your profiles with optional domains, categories and colors

image

```
from torch.cuda.nvtx import range_push, range_pop, range, range_start, range_end
import time

range_push('default domain')
time.sleep(.1)
range_pop()

range_push('custom domain', domain="mydomain")
time.sleep(.1)
range_pop(domain="mydomain")

range_push('custom category', category="mycategory")
time.sleep(.1)
range_pop()

range_push('custom category and domain', category="mycategory", domain="mydomain")
time.sleep(.1)
range_pop(domain="mydomain")

range_push('color', color=0xdd0000)
time.sleep(.1)
range_pop()

range_push('nonevalues', domain=None, category=None, color=None)
time.sleep(.1)
range_pop(domain=None)

id = range_start('range start', domain="mydomain", category="mycategory", color=0xeeeeee)
time.sleep(.1)
range_end(id, domain="mydomain")


@range('wrap', domain="mydomain", category="mycategory", color=0x0000ee)
def sleep():
    time.sleep(.1)


sleep()
```

cc @robieta @chaekit @aaronenyeshi @ngimel @nbcsm @guotuofeng @guyang3532 @gaoteng-git @tiffzhaofb

### CODE REVIEW

**Comment by milesial** [comment]
Not sure why the windows build fails because of undeclared symbol, does anybody know if the NVTX header is included differently in windows vs linus?

**Comment by milesial** [comment]
@malfet Do you have some guidance about the windows build fail?

**Comment by milesial** [comment]
@malfet gentle ping

File: torch/csrc/cuda/shared/nvtx.cpp [hunk 1]
```diff
@@ -1,22 +1,206 @@
...
 namespace torch {
 namespace cuda {
 namespace shared {
 
+#ifndef USE_ROCM
+
+// registry idea from NVIDIA/nvtx-plugins  @jdekhtiar
+class DomainCategoryRegistry {
+ public:
+  DomainCategoryRegistry() {}
```
Comment (malfet) [inline torch/csrc/cuda/shared/nvtx.cpp:17]: Please use modern C++
```suggestion
  DomainCategoryRegistry() = default;
```

File: torch/csrc/cuda/shared/nvtx.cpp [hunk 2]
```diff
@@ -1,22 +1,206 @@
...
+
+    nvtxDomainHandle_t domain_handle = nvtxDomainCreateA(domain_name.c_str());
+    domains[domain_name] = domain_handle;
+    return domain_handle;
+  }
+
+  uint64_t getCategory(
+      const std::string& category_name,
+      std::optional<std::string> domain_name) {
+    std::string id = category_name + "@" + domain_name.value_or("");
```
Comment (malfet) [inline torch/csrc/cuda/shared/nvtx.cpp:37]:
```suggestion
    const auto id = category_name + "@" + domain_name.value_or("");
```

File: torch/csrc/cuda/shared/nvtx.cpp [hunk 3]
```diff
@@ -1,22 +1,206 @@
...
+#ifndef USE_ROCM
+
+// registry idea from NVIDIA/nvtx-plugins  @jdekhtiar
+class DomainCategoryRegistry {
+ public:
+  DomainCategoryRegistry() {}
+
+  ~DomainCategoryRegistry() {
+    for (auto& domain : domains)
+      nvtxDomainDestroy(domain.second);
```
Comment (malfet) [inline torch/csrc/cuda/shared/nvtx.cpp:21]: Wrap even single line for/if body in curly braces
```suggestion
    for (auto& domain : domains) {
      nvtxDomainDestroy(domain.second);
    }
```

File: torch/csrc/cuda/shared/nvtx.cpp [hunk 4]
```diff
@@ -1,22 +1,206 @@
...
+  }
+
+  nvtxDomainHandle_t getDomain(const std::string& domain_name) {
+    auto it = domains.find(domain_name);
+    if (it != domains.end())
+      return it->second;
```
Comment (malfet) [inline torch/csrc/cuda/shared/nvtx.cpp:27]: PyTorch is C++17 project now, so let's use some of its functions
Also, why not const auto? You don't plan to modify it, do you?
```suggestion
    if ( const auto it = domains.find(domain_name); it != domains.end()) {
      return it->second;
   }   
```

File: torch/csrc/cuda/shared/nvtx.cpp [hunk 5]
```diff
@@ -1,22 +1,206 @@
...
+
+    nvtxDomainHandle_t domain_handle = nvtxDomainCreateA(domain_name.c_str());
```
Comment (malfet) [inline torch/csrc/cuda/shared/nvtx.cpp:29]:
```suggestion
    auto domain_handle = nvtxDomainCreateA(domain_name.c_str());
```

File: torch/csrc/cuda/shared/nvtx.cpp [hunk 6]
```diff
@@ -1,22 +1,206 @@
...
+    if (it != categories.end())
+      return it->second;
+
+    if (domain_name) // category local to domain
+      nvtxDomainNameCategoryA(
+          getDomain(domain_name.value()),
+          category_offset,
+          category_name.c_str());
+    else // category local to process
+      nvtxNameCategoryA(category_offset, category_name.c_str());
```
Comment (malfet) [inline torch/csrc/cuda/shared/nvtx.cpp:48]:
```suggestion
    if (domain_name) {
      // category local to domain
      nvtxDomainNameCategoryA(
          getDomain(domain_name.value()),
          category_offset,
          category_name.c_str());
    } else{ 
     // category local to process
      nvtxNameCategoryA(category_offset, category_name.c_str());
    }
```

File: torch/csrc/cuda/shared/nvtx.cpp [hunk 7]
```diff
@@ -1,22 +1,206 @@
...
+    const std::string& msg,
+    std::optional<std::string> domain,
+    std::optional<std::string> category,
+    std::optional<uint32_t> color) {
+  nvtxEventAttributes_t eventAttrib = {0};
+  eventAttrib.version = NVTX_VERSION;
+  eventAttrib.size = NVTX_EVENT_ATTRIB_STRUCT_SIZE;
+
+  if (category)
+    eventAttrib.category = registry.getCategory(category.value(), domain);
```
Comment (malfet) [inline torch/csrc/cuda/shared/nvtx.cpp:72]:
```suggestion
  if (category) {
    eventAttrib.category = registry.getCategory(category.value(), domain);
  }
```

File: torch/csrc/cuda/shared/nvtx.cpp [hunk 8]
```diff
@@ -1,22 +1,206 @@
...
+int rangePush(
+    const std::string& msg,
+    std::optional<std::string> domain,
+    std::optional<std::string> category,
+    std::optional<uint32_t> color) {
+#ifdef USE_ROCM
+  return roctxRangePushA(msg.c_str());
+#else
+  if (!domain && !category && !color)
+    return nvtxRangePushA(msg.c_str());
```
Comment (malfet) [inline torch/csrc/cuda/shared/nvtx.cpp:95]:
```suggestion
  if (!domain && !category && !color) {
    return nvtxRangePushA(msg.c_str());
  }
```

**Comment by malfet** [comment]
> @malfet Do you have some guidance about the windows build fail?

A bit confused why I was assigned a reviewer, but sure. Build fails with the following errors:
```
C:\actions-runner\_work\pytorch\pytorch\torch\csrc\cuda\shared\nvtx.cpp(24): error C3646: 'getDomain': unknown override specifier
C:\actions-runner\_work\pytorch\pytorch\torch\csrc\cuda\shared\nvtx.cpp(24): error C2059: syntax error: '('
C:\actions-runner\_work\pytorch\pytorch\torch\csrc\cuda\shared\nvtx.cpp(24): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
```
which suggests that `nvToolsExt.a` is not getting installed on Windows, which seems to correlate with
https://github.com/pytorch/test-infra/blob/61d67bfa403d61b0548ff6290f270c6b5d31462f/aws/ami/windows/scripts/Installers/Install-CUDA-Tools.ps1#L13

**Comment by milesial** [comment]
Ah yes NVTX1 on windows instead of NVTX2 must explain it, thanks!
Seems like the best way to handle it is to wait for NVTX3 to get merged in https://github.com/pytorch/pytorch/pull/90689

Not sure about the reviewer assignment, I see @ bdhirsh assigned you during triage so I assumed you were in charge of that part of the code.

I'll apply your suggestions tomorrow.

File: torch/csrc/cuda/shared/nvtx.cpp (hunk 1 above)
Comment (milesial) [reply torch/csrc/cuda/shared/nvtx.cpp]: Fixed in 30804cc

File: torch/csrc/cuda/shared/nvtx.cpp (hunk 2 above)
Comment (milesial) [reply torch/csrc/cuda/shared/nvtx.cpp]: Fixed in 30804cc

File: torch/csrc/cuda/shared/nvtx.cpp (hunk 3 above)
Comment (milesial) [reply torch/csrc/cuda/shared/nvtx.cpp]: Fixed in 30804cc

File: torch/csrc/cuda/shared/nvtx.cpp (hunk 4 above)
Comment (milesial) [reply torch/csrc/cuda/shared/nvtx.cpp]: Right, I use more consts in 30804cc

File: torch/csrc/cuda/shared/nvtx.cpp (hunk 5 above)
Comment (milesial) [reply torch/csrc/cuda/shared/nvtx.cpp]: Fixed in 30804cc

File: torch/csrc/cuda/shared/nvtx.cpp (hunk 6 above)
Comment (milesial) [reply torch/csrc/cuda/shared/nvtx.cpp]: Fixed in 30804cc

File: torch/csrc/cuda/shared/nvtx.cpp (hunk 7 above)
Comment (milesial) [reply torch/csrc/cuda/shared/nvtx.cpp]: Fixed in 30804cc

File: torch/csrc/cuda/shared/nvtx.cpp (hunk 8 above)
Comment (milesial) [reply torch/csrc/cuda/shared/nvtx.cpp]: Fixed in 30804cc

**Comment by malfet** [comment]
@milesial did you close because NVTX3 was finally integrated everywhere? Can you please confirm that it works as expected for you?

**Comment by milesial** [comment]
Oops, forgot about this PR and deleted my fork...
I'll reopen. As far as I know we're still waiting for https://github.com/pytorch/pytorch/pull/97582 to reland for NVTX3 to be integrated, and then the features in this PR can be merged.

**Pull request closed without merging**